snuba/datasets/events.py

```diff
@@ -162,7 +163,13 @@ def __init__(self):
             sample_expr=sample_expr)
 
         super(EventsDataSet, self).__init__(
-            schema
+            schema=schema,
+            processor=EventsProcessor(promoted_tag_columns),
+            message_topic="events",
+            message_consumer_group="snuba-consumers",
+            forward_topic="event-replacements",
+            forward_consumer_group="snuba-replacers",
```

Comment:
I cannot think about a specific case (and I am ok in renaming back to replacer). The root cause of the issue is that the events consumer can do two things out of a message: process and store in clickhouse or put it on another Kafka topic, which means the consumer needs to have access to Kafka because it can produce as well. 
If we took a step back and make the dataset implementation more generic letting the dataset author provide their own consumer class, the consumer script would have to instantiate this passing what is needed to instantiate the Kafka consumer which is very ad-hoc for events. This assuming we want to reuse the consumer.py script.

To me the right way of doing this would have been to handle replacements and events in two separate topics and simplify the consumer so that it would only process and store in clickhouse. But this requires refactoring on Sentry side as well. 
What do you think ?

Reply:
Changed back to replacement

Reply:
> If we took a step back and make the dataset implementation more generic letting the dataset author provide their own consumer class, the consumer script would have to instantiate this passing what is needed to instantiate the Kafka consumer which is very ad-hoc for events. This assuming we want to reuse the consumer.py script.

I'm not sure that I understand this specific comment. (I also haven't been able to put enough time into thinking about how these pieces fit together to have a constructive and/or concrete opinion about how we should move forward.) Overall, I think as long as we're on the same page about trying to find a way to split replacements out of the generic dataset at a later time (and also making the replacer concepts not required for new datasets in the meantime), that's good enough for me right now, and we can sort out the specific implementation details later.

> To me the right way of doing this would have been to handle replacements and events in two separate topics and simplify the consumer so that it would only process and store in clickhouse. But this requires refactoring on Sentry side as well.

There's a causal relationship between the two classes of events, unfortunately. The replacement events (merge, delete, etc.) need to happen after the events that are being mutated are inserted, otherwise we could end up "deleting" events before they are ever written (making it so the deletion never actually has the intended effect.)

Reply:
I forgot we used the replacer for deleting events as well.

Reply:
And yes, setting the replacement topic to None and writing a processor that does not return "replace" will be enough to have a dataset without replacer.